File: LeetCode/container_with_most_water.js
/*
You are given an integer array height of length n. There are n vertical lines drawn such that the two endpoints of the ith line are (i, 0) and (i, height[i]).

Find two lines that together with the x-axis form a container, such that the container contains the most water.

Return the maximum amount of water a container can store.


* @param {number[]} height
 * @return {number}
 */
let maxArea = function (height) {
  let right = height.length - 1,
    left = 0,
    maxarea = 0;
  while (left < right) {
    let width = right - left;
    maxarea = Math.max(maxarea, Math.min(height[left], height[right]) * width);
    if (height[left] <= height[right]) {
      left++;
    } else {
      right--;
    }
  }

  return maxarea;
};

File: LeetCode/Container with most water.cpp
//timelimit exceeded
class Solution {
public:
    int maxArea(vector<int>& height) {
        long long int m=0;
        for(long long int i=0;i<height.size()-1;i++)
        {
            long long int j=i+1;
            while(j>i && j<height.size()){
            long long int a=min(height[i],height[j]);
            long long int b=abs((j+1)-(i+1));
            m=max(m,a*b);
            j++;
         }
        }
        return m;
    }
};

//optimal solution
class Solution {
public:
    int maxArea(vector<int>& height) {
        long long int m=0,i=0,j=height.size()-1;
        while(i<j)
        {
           long long int a=abs(i-j);
           long long int b=min(height[i],height[j]);
            m=max(m,a*b);
            if(height[i]<height[j]) i++;
            else if(height[i]>height[j]) j--;
            else
            {
                i++;
                j--;
            }
        }
        return m;
    }
};
